braindecode/regressor.py

```diff
@@ -0,0 +1,194 @@
+        if not training:
+            cbs = self._default_callbacks + self.callbacks
+            epoch_cbs = []
+            for name, cb in cbs:
+                if (cb.__class__.__name__ == 'CroppedTrialEpochScoring') and (
+                    hasattr(cb, 'supercrop_inds_')) and (cb.on_train == False):
+                    epoch_cbs.append(cb)
+            # for trialwise decoding stuffs it might also be we don't have
+            # cropped loader, so no indices there
+            if len(epoch_cbs) > 0:
+                assert hasattr(self, '_last_supercrop_inds')
+                for cb in epoch_cbs:
+                    cb.supercrop_inds_.append(self._last_supercrop_inds)
+                del self._last_supercrop_inds
+
+    def predict_with_supercrop_inds_and_ys(self, dataset):
+        preds = []
+        i_supercrop_in_trials = []
+        i_supercrop_stops = []
+        supercrop_ys = []
+        for X, y, i in self.get_iterator(dataset, drop_index=False):
+            i_supercrop_in_trials.append(i[0].cpu().numpy())
+            i_supercrop_stops.append(i[2].cpu().numpy())
+            preds.append(self.predict_proba(X))
+            supercrop_ys.append(y.cpu().numpy())
+        preds = np.concatenate(preds)
+        i_supercrop_in_trials = np.concatenate(i_supercrop_in_trials)
+        i_supercrop_stops = np.concatenate(i_supercrop_stops)
+        supercrop_ys = np.concatenate(supercrop_ys)
+        return dict(
+            preds=preds, i_supercrop_in_trials=i_supercrop_in_trials,
+            i_supercrop_stops=i_supercrop_stops, supercrop_ys=supercrop_ys)
+
+    # Removes default EpochScoring callback computing 'accuracy' to work properly
+    # with cropped decoding.
+    @property
+    def _default_callbacks(self):
+        return [
+            ("epoch_timer", EpochTimer()),
+            (
+                "train_loss",
+                BatchScoring(
+                    train_loss_score,
+                    name="train_loss",
+                    on_train=True,
+                    target_extractor=noop,
+                ),
+            ),
+            (
+                "valid_loss",
+                BatchScoring(
+                    valid_loss_score, name="valid_loss", target_extractor=noop,
+                ),
+            ),
+            ("print_log", PrintLog()),
+        ]
+
+
+class ThrowAwayIndexLoader(object):
+    def __init__(self, net, loader):
+        self.net = net
+        self.loader = loader
+        self.last_i = None
+
+    def __iter__(self, ):
+        normal_iter = self.loader.__iter__()
+        for batch in normal_iter:
+            if len(batch) == 3:
+                x,y,i = batch
+                # Store for scoring callbacks
+                self.net._last_supercrop_inds = i
+            else:
+                x,y = batch
+
+            # TODO: should be on dataset side
+            if hasattr(x, 'type'):
+                x = x.type(torch.float32)
+                y = y.type(torch.float32)
```

Comment:
Yes, it could definitely be merged. As also the 'TODO' message in the code says, it would be favorable that the datasets already give data of the appropriate type. This would remove the necessity of casting in the <code>ThrowAwayIndexLoader</code>.  This would mean that the datasets need to know which task ithey are being used for.

Reply:
For now would agree with @sliwy  solution to introduce a flag and just have one class

Reply:
I'll introduce a flag <code>is_regression</code> to the <code>ThrowAwayIndexLoader</code>. I will then move this class to <code>braindecode.util</code> and import it in <code>classifier.py</code> and <code>regressor.py</code>, OK?